libcxx/test/tools/clang_tidy_checks/header_exportable_declarations.cpp

```diff
@@ -199,50 +223,121 @@ static bool is_viable_declaration(const clang::NamedDecl* decl) {
+  if (decl.getDeclContext()->isNamespace())
+    return true;
+
+  if (is_valid_global_name(name))
+    return true;
+
+  return file_type != header_exportable_declarations::FileType::Header;
+}
+
+static bool is_module(header_exportable_declarations::FileType file_type) {
+  switch (file_type) {
+  case header_exportable_declarations::FileType::Module:
+  case header_exportable_declarations::FileType::ModulePartition:
+  case header_exportable_declarations::FileType::CompatModule:
+  case header_exportable_declarations::FileType::CompatModulePartition:
+    return true;
+
+  case header_exportable_declarations::FileType::Header:
+  case header_exportable_declarations::FileType::CHeader:
+    return false;
+
+  case header_exportable_declarations::FileType::Unknown:
+    llvm::errs() << "This should be unreachable.\n";
+    break;
+  }
+}
+
 void header_exportable_declarations::check(const clang::ast_matchers::MatchFinder::MatchResult& result) {
   if (const auto* decl = result.Nodes.getNodeAs<clang::NamedDecl>("header_exportable_declarations"); decl != nullptr) {
     if (!is_viable_declaration(decl))
       return;
 
     std::string name = get_qualified_name(*decl);
+    if (name.empty())
+      return;
+
     if (is_reserved_name(name))
       return;
 
     // For modules only take the declarations exported.
-    if (file_type_ == FileType::ModulePartition || file_type_ == FileType::Module)
+    if (is_module(file_type_))
       if (decl->getModuleOwnershipKind() != clang::Decl::ModuleOwnershipKind::VisibleWhenImported)
         return;
 
+    if (!is_valid_declaration_context(*decl, name, file_type_))
+      return;
+
     if (decls_.contains(name)) {
       // For modules avoid exporting the same named declaration twice. For
       // header files this is common and valid.
-      if (file_type_ == FileType::ModulePartition)
+      if (file_type_ == FileType::ModulePartition || file_type_ == FileType::CompatModulePartition)
         // After the warning the script continues.
         // The test will fail since modules have duplicated entries and headers not.
         llvm::errs() << "Duplicated export of '" << name << "'.\n";
       else
         return;
     }
 
-    std::cout << "using " << std::string{name} << ";\n";
+    // For named declarations in std this is valid
+    //   using std::foo;
+    // for named declarations it is invalid to use
+    //   using bar;
+    // Since fully qualifying named declarations in the std namespace is valid
+    // using fully qualified names unconditionally.
+    std::cout << "using ::" << std::string{name} << ";\n";
     decls_.insert(name);
+  } else if (const auto* decl = result.Nodes.getNodeAs<clang::NamedDecl>("cheader_exportable_declarations");
+             decl != nullptr) {
+    if (decl->getDeclContext()->isNamespace())
+      return;
+
+    if (!is_viable_declaration(decl))
+      return;
+
+    std::string name = get_qualified_name(*decl);
+    if (is_reserved_name(name))
```

Comment:
Do we see the declaration of `::_Exit` here? If so do we filter reserved names correctly? I think the result is that right now we're not generating a test for std.compat's definition of `::_Exit`.

Reply:
We indeed didn't filter `::_Exit` properly, this has been fixed.